File: python/src/kalourmade_sailup/http.py
from typing import Optional, Protocol

import httpx

from .exceptions import NetworkError


class Transport(Protocol):
    def request(
        self,
        method: str,
        path: str,
        query: Optional[dict] = None,
        json_body: Optional[dict] = None,
    ) -> tuple[int, str]:
        ...


class HttpxTransport:
    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.sailup.io/v1",
        timeout: float = 30.0,
    ):
        self._client = httpx.Client(
            base_url=base_url,
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=timeout,
        )

    def request(
        self,
        method: str,
        path: str,
        query: Optional[dict] = None,
        json_body: Optional[dict] = None,
    ) -> tuple[int, str]:
        try:
            response = self._client.request(method, path, params=query, json=json_body)
        except httpx.HTTPError as exc:
            raise NetworkError(str(exc)) from exc

        return response.status_code, response.text

File: node/src/http.ts
import { NetworkError } from './errors.js';

export interface HttpResponse {
  status: number;
  body: string;
}

export interface Transport {
  request(
    method: string,
    path: string,
    query?: Record<string, string | number>,
    jsonBody?: unknown
  ): Promise<HttpResponse>;
}

export class FetchTransport implements Transport {
  constructor(
    private readonly apiKey: string,
    private readonly baseUrl: string = 'https://api.sailup.io/v1',
    private readonly timeoutMs: number = 30000
  ) {}

  async request(
    method: string,
    path: string,
    query?: Record<string, string | number>,
    jsonBody?: unknown
  ): Promise<HttpResponse> {
    const url = new URL(this.baseUrl.replace(/\/$/, '') + path);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        url.searchParams.set(key, String(value));
      }
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const response = await fetch(url, {
        method,
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: jsonBody !== undefined ? JSON.stringify(jsonBody) : undefined,
        signal: controller.signal,
      });
      const body = await response.text();
      return { status: response.status, body };
    } catch (err) {
      throw new NetworkError(err instanceof Error ? err.message : String(err));
    } finally {
      clearTimeout(timer);
    }
  }
}
